### PR TITLE
libs/log: format []byte as base 16

### LINKED ISSUE #5806: libs/log: format []byte as base 16 (upper case)
Triggered by https://github.com/tendermint/tendermint/pull/5805

We should modify our logger to format byte slices as base 16 (upper case) automatically, which will remove the need for `fmt.Sprintf("%X", hash)` calls.

### PR DESCRIPTION
**Pull request by laniehei**
## Description

- Formats []byte type as base 16 in tm_logger
- Removes need for hash2str function

Closes: #5806

### CODE REVIEW

**Review by alexanderbez — APPROVED** [review]

**Comment by tessr** [comment]
Hi @laniehei, thank you so much for this contribution! It looks great. 

Before we merge this, can you fix the spelling mistakes caught by the linter ("fourty" vs "forty"). Thank you!

**Review by erikgrinaker — CHANGES_REQUESTED** [review]
Thanks for the PR! This isn't spelled out in the issue, but some of the motivation for this (as noted in #5805) is to avoid the overhead of hex-encoding when the message won't get logged. For example, we do fairly frequent debug logging of message payloads in the P2P networking layer, and don't want to spend CPU cycles doing this string formatting when debug-logging isn't enabled.

I haven't looked into the logging infrastructure, but would it be possible to do this string formatting _after_  the level filtering, or alternatively skipping it?

File: libs/log/tm_logger.go
Comment (dshulyak) [inline libs/log/tm_logger.go:60]: bytes.ToUpper doesn't encode as hexademical (base16). in general stdlib provides `encoding/hex` module, it might not encode in upper-case. also it is a bit wasteful to do it here.

if value implements either `encoding.TextMarshaler` or `fmt.Stringer` it will be marshalled accordingly to the implementation after passing log level check

Comment (laniehei) [reply libs/log/tm_logger.go]: Hi @dshulyak, thank you for the feedback! For my own knowledge, why might the encoding/hex module might not encode in uppercase?

**Comment by laniehei** [comment]
@erikgrinaker thank you for the clarification on the motivation behind the issue! It's an interesting usecase. I think that I've implemented the Stringer interface in the way that @dshulyak was nice enough to explain to me. Does this look to be more in line with what @dshulyak explained, and what you're looking for? Thanks for your feedback!

File: libs/log/tm_logger.go
Comment (dshulyak) [reply libs/log/tm_logger.go]: library is simply using lowercase [chars](https://github.com/golang/go/blob/master/src/encoding/hex/hex.go#L15) for encoding. i guess it is a common preference

File: libs/log/tm_logger.go
Comment (dshulyak) [inline libs/log/tm_logger.go:62]: the common approach is to specify type outside of the logger. if it is possible block.Hash() should return Hash type, but if it introduces too many changes, maybe just convert it in place (where hash2str was used previously). better to ask maintainers though, maybe this was the original idea

File: libs/log/logger.go
Comment (erikgrinaker) [reply libs/log/logger.go:21]: Forgot to mention that we actually already have a type `bytes.HexBytes` that does this, so we can just use that:

https://github.com/tendermint/tendermint/blob/8e6194626eb60aef79052cfa93949bdc61e6ce05/libs/bytes/bytes.go#L12

File: libs/log/tm_logger.go
Comment (erikgrinaker) [reply libs/log/tm_logger.go:62]: It isn't really viable to enforce the use of `HexBytes` instead of `[]byte` everywhere, particularly in cases like external libraries and generated code (e.g. Protobuf). I prefer this to be handled automagically by the logging layer, as it is entirely a logging concern.

Comment (dshulyak) [reply libs/log/tm_logger.go:62]: there were like 10 places where hash2str was logged) in general yes *everywhere* is hard. but it is viable to have a consistent textual representation for hashes across the application, for instance gogo codec supports casttype directive. json uses base64 for raw bytes by default, so it must be aware of the type as well if it is used at any point

Comment (erikgrinaker) [reply libs/log/tm_logger.go:62]: This isn't just about hashes, it's about everywhere we log byte slices. We want to avoid having raw binary data output to a terminal, since that's not very readable and also can mess up the terminal with escape codes and other special characters.

Comment (erikgrinaker) [reply libs/log/tm_logger.go:62]: You do bring up a good point about consistent formatting though, so the logger shouldn't override the formatting if the input type already implements `fmt.Stringer` itself.

Comment (dshulyak) [reply libs/log/tm_logger.go:62]: there will be inconsistencies such as in [TX API](https://docs.tendermint.com/master/rpc/#/Info/tx) data returned as base64, not hex, that won't be tracked automatically

Comment (erikgrinaker) [reply libs/log/tm_logger.go:62]: Sure, data that should have a canonical textual representation should implement `fmt.Stringer` and any other relevant interfaces. We can deal with that separately, but that's not what this issue/PR is about -- this is about avoiding writing raw binary data to logs, which we e.g. do with debug logging in the P2P networking layer where we deal with raw byte slices all over the place.

File: libs/bytes/bytes_test.go
Comment (tessr) [inline libs/bytes/bytes_test.go:70]: nit: spurious newline

File: libs/bytes/bytes_test.go
Comment (melekes) [inline libs/bytes/bytes_test.go:70]:
```suggestion
	if _, err := strconv.ParseInt(hs.String(), 16, 64); err != nil {
```

File: libs/bytes/bytes_test.go
Comment (melekes) [inline libs/bytes/bytes_test.go:73]:
```suggestion
```

File: libs/log/tm_logger.go
Comment (melekes) [inline libs/log/tm_logger.go:57]: maybe we should move this to https://github.com/tendermint/tendermint/blob/09cf0bcb012d37de33b41425d82a77bbb31ed1b5/libs/log/tmfmt_logger.go#L107

that way we'll eliminate duplication

**Review by melekes — COMMENTED** [review]
thanks for working on this 👍 I think we can simplify the code a little bit, but other than that it looks good

File: libs/log/tm_logger.go
Comment (laniehei) [reply libs/log/tm_logger.go:57]: Does tmfmt_logger wrap tm_logger? Maybe I misunderstand how these loggers are used.

Comment (melekes) [reply libs/log/tm_logger.go:57]: I believe `tm_logger` wraps `tmfmt_logger`

Comment (laniehei) [reply libs/log/tm_logger.go:57]: Wouldn't we want to convert this to hexbytes here in tm_logger, because  and it's unnecessary for srcLogger to handle the conversion/have knowledge of the hexByte conversion? Seems like if we switched out the implementation of srcLogger at some future date, this may cause issues in the future (or failure to move the hexbytes conversion into this new implementation), right? 

What is the benefit of moving this to the other logger?

File: libs/bytes/bytes_test.go
Comment (alexanderbez) [inline libs/bytes/bytes_test.go:72]:
```suggestion
	require.NoError(t, strconv.ParseInt(hs.String(), 16, 64))
```

File: libs/log/tm_logger.go
Comment (jesselang) [reply libs/log/tm_logger.go:57]: I think I found where `tm_logger` wraps `tmfmt_logger`: https://github.com/tendermint/tendermint/blob/master/libs/log/tm_logger.go#L43

I'm not familiar with the intended architecture design of the loggers, but @laniehei brings up some good questions about how future changes to the logging design could cause unintended regressions in this specialized formatting. How could we protect against this type of regression?

File: libs/log/tm_logger.go
Comment (jesselang) [reply libs/log/tm_logger.go]: Is it important to deal with that inconsistency before merging? Presumably, the code could do a type assertion for `fmt.Stringer` and use its function before resorting to the proposed default representation of base16?

Comment (erikgrinaker) [reply libs/log/tm_logger.go]: Yeah, I think that'd be great -- for all I know, we may already have `[]byte` types that implement `fmt.Stringer`. It should be sufficient to simply skip the hex encoding if the type implements `fmt.Stringer`, since the logger would presumably make use of the interface.

Comment (dshulyak) [reply libs/log/tm_logger.go]: `val.([]byte)` type check already returns false if value inside of the interface is not a []byte

Comment (erikgrinaker) [reply libs/log/tm_logger.go]: Well yes. But if it's a `[]byte` that also implements `fmt.Stringer` then we shouldn't override the formatting.

Comment (dshulyak) [reply libs/log/tm_logger.go]: i mean in order to implement fmt.Stringer you need to declare a new type in golang

Comment (erikgrinaker) [reply libs/log/tm_logger.go]: Ah, you're right! I thought this bit would work:

```golang
type Foo []byte

func (foo Foo) String() string {
	return "foo"
}

func main() {
	var foo interface{}
	foo = Foo{}
	if _, ok := foo.([]byte); ok {
		fmt.Println("is []byte")
	}
	if _, ok := foo.(fmt.Stringer); ok {
		fmt.Println("implements fmt.Stringer")
	}
}
```

But turns out the `[]byte` assertion fails. Which I suppose makes sense, but I though maybe it'd downcast the type in these cases. Thanks for correcting me!

Nevermind then, the current code is fine. Although I'd be curious to know how to handle this for all `[]byte`-derived types.

**Review by erikgrinaker — APPROVED** [review]

File: libs/log/tm_logger.go
Comment (dshulyak) [reply libs/log/tm_logger.go]: i think the simplest way is to use reflection, e.g. this snippet will assert if data is stored internally as []byte:

```go
func isByteSlice(val interface{}) bool {
	slice := reflect.TypeOf(val)
	if slice.Kind() != reflect.Slice {
		return false
	}
	return slice.Elem().Kind() == reflect.Uint8 // no kind for a byte
}
``` 

i would expect it to introduce certain overhead if used in the logger, needs to be doublechecked

Comment (erikgrinaker) [reply libs/log/tm_logger.go]: Yeah, reflection seems a bit excessive here.

File: libs/bytes/bytes_test.go
Comment (laniehei) [reply libs/bytes/bytes_test.go:72]: I don't follow how this would be a useful addition, because it pulls in an extra library from the third party testify, and this check is already achieved by checking whether the err returned from ParseInt is nil. Also, the suggested code has a bug in it, because ParseInt returns two values.

**Comment by tessr** [comment]
@marbar3778 this is failing on CI because the Docker username and pw are missing - did we change that recently and/or do I need to add new secrets to this repo?

**Comment by marbar3778** [comment]
> @marbar3778 this is failing Actions CI because the Docker username and pw are missing - did we change that recently and/or do I need to add new secrets to this repo?

When a contribution comes from a fork it CI will always fail where we are injecting secrets. This is a GitHub security measure.

File: libs/log/tm_logger.go
Comment (melekes) [reply libs/log/tm_logger.go:57]: > What is the benefit of moving this to the other logger?

de duplication of the code. we won't have to repeat the same logic 3 times (Info, Debug, Error).

> Wouldn't we want to convert this to hexbytes here in tm_logger, because and it's unnecessary for srcLogger to handle the conversion/have knowledge of the hexByte conversion?

no. if you look at the tmfmt_logger, you will see that it already has deep knowledge of the structure and does formatting (what we're trying to do here as well).

**Comment by melekes** [comment]
Closing in favor of https://github.com/tendermint/tendermint/pull/5960. I will make some minor changes and ensure this gets merged. Thanks @laniehei for your work 👍 🚀

**Comment by tessr** [comment]
Ah, that's too bad. Kind of a bummer that contributors can't get their own PRs merged. @laniehei, thank you for this contribution, and for sticking through several rounds of review!

> When a contribution comes from a fork it CI will always fail where we are injecting secrets. This is a GitHub security measure. 

I wonder if there's a way to work around this. It seems weird that you can't pass CI on PRs from forks--that might be when you want forks _the most_. I'll ask GitHub.

**Comment by laniehei** [comment]
@tessr happy to have contributed, and engaged with you all! Looking forward to what Github has to say, as I'd like to get PRs merged in the future!

**Pull request closed without merging**
